File: src/main/java/com/mozzartbet/hackaton/connect4/Michigan/Bot.java
package com.mozzartbet.hackaton.connect4.Michigan;

import com.mozzartbet.hackaton.connect4.model.GameBoard;
import com.mozzartbet.hackaton.connect4.model.GameConsts;
import com.mozzartbet.hackaton.connect4.model.Player;

import java.util.*;

public class Bot extends Player {

    protected final int HUMAN = 1;
    protected final int BOT = 2;
    private GameBoard gameBoard = new GameBoard();

    private boolean isMoveValid(int col){
        return gameBoard.getBoard()[0][col] == 0;
    }


    private static void dropPiece(int[][] board, int col, int player) {
        for (int row = GameConsts.ROWS - 1; row >= 0; row--) {
            if (board[row][col] == 0) {
                board[row][col] = player;
                return;
            }
        }
    }
    private boolean hasHorizontalSeries(int[][] board, int player) {
        for (int row = 0; row < GameConsts.ROWS; row++) {
            for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
                if (board[row][col] == player &&
                        board[row][col + 1] == player &&
                        board[row][col + 2] == player &&
                        board[row][col + 3] == player) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean hasVerticalSeries(int[][] board, int player) {
        for (int col = 0; col < GameConsts.COLUMNS; col++) {
            for (int row = 0; row < GameConsts.ROWS - 3; row++) {
                if (board[row][col] == player &&
                        board[row + 1][col] == player &&
                        board[row + 2][col] == player &&
                        board[row + 3][col] == player) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean hasDiagonalSeries(int[][] board, int player) {
        // Check positive diagonal
        for (int row = 0; row < GameConsts.ROWS - 3; row++) {
            for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
                if (board[row][col] == player &&
                        board[row + 1][col + 1] == player &&
                        board[row + 2][col + 2] == player &&
                        board[row + 3][col + 3] == player) {
                    return true;
                }
            }
        }

        // Check negative diagonal
        for (int row = 3; row < GameConsts.ROWS; row++) {
            for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
                if (board[row][col] == player &&
                        board[row - 1][col + 1] == player &&
                        board[row - 2][col + 2] == player &&
                        board[row - 3][col + 3] == player) {
                    return true;
                }
            }
        }

        return false;
    }

    private List<int[][]> getNextStates(int[][] board, int player) {
        List<int[][]> states = new ArrayList<>();
        for (int col = 0; col < GameConsts.COLUMNS ; col++) {
            if (isMoveValid(col)) {
                int[][] nextBoard = copyBoard(board);
                dropPiece(nextBoard, col, player);
                states.add(nextBoard);
            }
        }
        return states;
    }

    private int[][] copyBoard(int[][] board) {
        int[][] copy = new int[GameConsts.ROWS][GameConsts.COLUMNS];
        for (int i = 0; i < GameConsts.ROWS; i++) {
            copy[i] = Arrays.copyOf(board[i], GameConsts.COLUMNS);
        }
        return copy;
    }

    private static int getNextOpenRow(int[][] board, int col) {
        for (int row = GameConsts.ROWS - 1; row >= 0; row--) {
            if (board[row][col] == 0) {
                return row;
            }
        }
        return -1; // Should not reach here if the move is valid
    }

    private int getBestMove(int player) {
        Set<Integer> validLocations = getValidLocations(gameBoard.getBoard());
        int bestScore = Integer.MIN_VALUE;
        Set<Integer> bestMoves = new HashSet<>();

        // Attempt to find the optimal move, avoiding opening in the corners
        Set<Integer> optimalMoves = new HashSet<>(validLocations);
        optimalMoves.removeAll(Arrays.asList(0, GameConsts.COLUMNS - 1));

        if (!optimalMoves.isEmpty()) {
            bestMoves.add(getRandomElement(optimalMoves));
        } else {
            bestMoves.add(getRandomElement(validLocations));
        }

        for (int col : validLocations) {
            int row = getNextOpenRow(gameBoard.getBoard(), col);
            int[][] tempBoard = copyBoard(gameBoard.getBoard());
            dropPiece(tempBoard, col, player);
            int score = min(tempBoard, Integer.MIN_VALUE, Integer.MAX_VALUE, 3, 0, 0, 0);

            if (score > bestScore) {
                bestMoves.clear();
                bestMoves.add(col);
                bestScore = score;
            } else if (score == bestScore) {
                bestMoves.add(col);
            }
        }

        // Random selection among the best moves
        return getRandomElement(bestMoves);
    }


    private int min(int[][] currentState, int alpha, int beta, int depth, int seriesDepth, int seriesCount, int winCount) {
        if (isTerminalNode(currentState) || depth == 0) {
            return evaluateBoard(currentState, HUMAN, depth, seriesDepth, seriesCount, winCount);
        }

        int minEval = Integer.MAX_VALUE;

        for (int[][] nextState : getNextStates(currentState, HUMAN)) {
            int evalVal = max(nextState, alpha, beta, depth - 1, seriesDepth, seriesCount, winCount);
            minEval = Math.min(minEval, evalVal);
            beta = Math.min(beta, evalVal);
            if (beta <= alpha) {
                break;
            }
        }

        return minEval;
    }

    private int max(int[][] currentState, int alpha, int beta, int depth, int seriesDepth, int seriesCount, int winCount) {
        if (isTerminalNode(currentState) || depth == 0) {
            return evaluateBoard(currentState, BOT, depth, seriesDepth, seriesCount, winCount);
        }

        int maxEval = Integer.MIN_VALUE;

        for (int[][] nextState : getNextStates(currentState, BOT)) {
            int evalVal = min(nextState, alpha, beta, depth - 1, seriesDepth, seriesCount, winCount);
            maxEval = Math.max(maxEval, evalVal);
            alpha = Math.max(alpha, evalVal);
            if (beta <= alpha) {
                break;
            }
        }

        return maxEval;
    }


    private int evaluateBoard(int[][] board, int player, int depth, int seriesDepth, int seriesCount, int winCount) {

        int score = 0;

        // Check for winning positions
        if (winningMove(board, HUMAN)) {
            return -1000;
        } else if (winningMove(board, BOT)) {
            return 1000;
        }

        score += depth;

        score += seriesCount * 10;

        score += seriesDepth;

        score += winCount * 100;

        score += evaluateClosing(board, player);

        // Heuristic 1: Prevent opponent's winning move
        for (int col : getValidLocations(board)) {
            int row = getNextOpenRow(board, col);
            int[][] tempBoard = copyBoard(board);
            dropPiece(tempBoard, col, player);
            if (winningMove(tempBoard, player)) {
                score -= 800;  // Priority on preventing opponent's winning move
            }
        }
        int[][] tempBoard = copyBoard(board);

        // Heuristic 2: Block opponent's horizontal series
        if (hasHorizontalSeries(board, HUMAN)) {
            score -= 500;  // Priority on blocking opponent's horizontal series
        }

        // Heuristic 3: Block opponent's vertical series
        if (hasVerticalSeries(board, HUMAN)) {
            score -= 1000;  // Priority on blocking opponent's vertical series
        }

        // Heuristic 4: Block opponent's diagonal series
        if (hasDiagonalSeries(board, HUMAN)) {
            score -= 300;  // Priority on blocking opponent's diagonal series
        }

        return score;
    }


    private int evaluateClosing(int[][] board, int player) {

        Set<Integer> validLocations = getValidLocations(board);

        // Heuristic 1: Winning moves
        for (int col : validLocations) {
            int row = getNextOpenRow(board, col);
            int[][] tempBoard = copyBoard(board);
            dropPiece(tempBoard, col, BOT);
            if (winningMove(tempBoard, HUMAN)) {
                return 1000;  // Priority on winning move
            }
        }

        // Heuristic 2: Blocking the opponent
        for (int col : validLocations) {
            int row = getNextOpenRow(board, col);
            int[][] tempBoard = copyBoard(board);
            dropPiece(tempBoard, col, BOT);
            if (winningMove(tempBoard, HUMAN)) {
                return 900;  // Priority on blocking the opponent
            }
        }

        // Heuristic 3: Attacks with more options
        int maxPlayerOptions = 0;
        for (int col : validLocations) {
            int row = getNextOpenRow(board, col);
            int[][] tempBoard = copyBoard(board);
            dropPiece(tempBoard, col, HUMAN);
            int playerOptions = getValidLocations(tempBoard).size();
            if (playerOptions > maxPlayerOptions) {
                maxPlayerOptions = playerOptions;
            }
        }

        // Heuristic 4: Blocking opponent's options
        int maxOpponentOptions = 0;
        for (int col : validLocations) {
            int row = getNextOpenRow(board, col);
            int[][] tempBoard = copyBoard(board);
            dropPiece(tempBoard, col, HUMAN);
            int opponentOptions = getValidLocations(tempBoard).size();
            if (opponentOptions > maxOpponentOptions) {
                maxOpponentOptions = opponentOptions;
            }
        }


        return maxPlayerOptions * 10 - maxOpponentOptions * 5;
    }


    private boolean isTerminalNode(int[][] board) {
        return winningMove(board, HUMAN) || winningMove(board, BOT) || getValidLocations(board).isEmpty();
    }

    private boolean winningMove(int[][] board, int player) {
        for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
            for (int row = 0; row < GameConsts.ROWS; row++) {
                if (board[row][col] == player &&
                        board[row][col + 1] == player &&
                        board[row][col + 2] == player &&
                        board[row][col + 3] == player) {
                    return true;
                }
            }
        }

        // Check vertical
        for (int col = 0; col < GameConsts.COLUMNS ; col++) {
            for (int row = 0; row < GameConsts.ROWS - 3; row++) {
                if (board[row][col] == player &&
                        board[row + 1][col] == player &&
                        board[row + 2][col] == player &&
                        board[row + 3][col] == player) {
                    return true;
                }
            }
        }

        // Check positive diagonal
        for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
            for (int row = 0; row < GameConsts.ROWS - 3; row++) {
                if (board[row][col] == player &&
                        board[row + 1][col + 1] == player &&
                        board[row + 2][col + 2] == player &&
                        board[row + 3][col + 3] == player) {
                    return true;
                }
            }
        }

        // Check negative diagonal
        for (int col = 0; col < GameConsts.COLUMNS - 3; col++) {
            for (int row = 3; row < GameConsts.ROWS; row++) {
                if (board[row][col] == player &&
                        board[row - 1][col + 1] == player &&
                        board[row - 2][col + 2] == player &&
                        board[row - 3][col + 3] == player) {
                    return true;
                }
            }
        }

        return false;
    }


    private Set<Integer> getValidLocations(int[][] board) {
        Set<Integer> validLocations = new HashSet<>();
        for (int col = 0; col < GameConsts.COLUMNS; col++) {
            if (isMoveValid(col)) {
                validLocations.add(col);
            }
        }
        return validLocations;
    }


    private static int getRandomElement(Set<Integer> set) {
        int size = set.size();
        int item = new Random().nextInt(size);
        int i = 0;
        for (int col : set) {
            if (i == item) {
                return col;
            }
            i++;
        }
        return 1; // Should not reach here
    }

    @Override
    public void configure(long timeoutMillis) {

    }

    @Override
    public void move() {
        move = getBestMove(BOT);
        gameBoard.placeCounter(move, BOT);
    }

    @Override
    public void stop() {

    }

    @Override
    public void opponentMove(int move) {
        gameBoard.placeCounter(move, HUMAN);
    }

    @Override
    public void finished(int winner) {

    }
}

File: bot.py
import numpy as np
import random

ROWS = 6
COLS = 8
EMPTY = 0
PLAYER_X = 1
PLAYER_O = 2

def create_board():
    return np.zeros((ROWS, COLS), dtype=int)

def is_valid_move(board, col):
    return board[0][col] == EMPTY

def drop_piece(board, col, player):
    for row in range(ROWS-1, -1, -1):
        if board[row][col] == EMPTY:
            board[row][col] = player
            return

def get_next_states(board, player):
    states = []
    for col in range(COLS):
        if is_valid_move(board, col):
            next_board = board.copy()
            drop_piece(next_board, col, player)
            states.append(next_board)
    return states


def evaluate_closing(board, player):
    opponent = PLAYER_X if player == PLAYER_O else PLAYER_O
    valid_locations = get_valid_locations(board)

    # Heuristika 1: Pobednički potezi
    for col in valid_locations:
        row = get_next_open_row(board, col)
        temp_board = board.copy()
        drop_piece(temp_board, col, player)
        if winning_move(temp_board, player):
            return 1000  # Prioritet na pobedničkom potezu

    # Heuristika 2: Blokiranje protivnika
    for col in valid_locations:
        row = get_next_open_row(board, col)
        temp_board = board.copy()
        drop_piece(temp_board, col, opponent)
        if winning_move(temp_board, opponent):
            return 900  # Prioritet na blokiranju protivnika

    # Heuristika 3: Napadi sa više opcija
    max_player_options = 0
    for col in valid_locations:
        row = get_next_open_row(board, col)
        temp_board = board.copy()
        drop_piece(temp_board, col, player)
        player_options = len(get_valid_locations(temp_board))
        if player_options > max_player_options:
            max_player_options = player_options

    # Heuristika 4: Blokiranje protivnikovih opcija
    max_opponent_options = 0
    for col in valid_locations:
        row = get_next_open_row(board, col)
        temp_board = board.copy()
        drop_piece(temp_board, col, opponent)
        opponent_options = len(get_valid_locations(temp_board))
        if opponent_options > max_opponent_options:
            max_opponent_options = opponent_options

    return max_player_options * 10 - max_opponent_options * 5

def evaluate_board(board, player, depth, series_depth, series_count, win_count):
    score = 0

    # Heuristic 1: Ranks a board based on whether it's a win, loss, or immediate opponent victory setup
    if winning_move(board, PLAYER_X):
        return -1000  # Opponent (PLAYER_X) wins
    elif winning_move(board, PLAYER_O):
        return 1000  # Bot (PLAYER_O) wins

    # Heuristic 2: Depth of the board within the MinMax tree
    score += depth

    # Heuristic 3: Number of series of three consecutive tokens
    score += series_count * 10

    # Heuristic 4: Consider the depths of the series of three tokens
    score += series_depth

    # Heuristic 5: Give value to the number of wins a board has
    score += win_count * 100

    # Dodajemo ocenu za zatvaranje
    score += evaluate_closing(board, player)

    return score

def is_terminal_node(board):
    return (winning_move(board, PLAYER_X) or 
            winning_move(board, PLAYER_O) or 
            len(get_valid_locations(board)) == 0)

def winning_move(board, piece):
    # Check horizontal
    for col in range(COLS - 3):
        for row in range(ROWS):
            if board[row][col] == piece and board[row][col+1] == piece and board[row][col+2] == piece and board[row][col+3] == piece:
                return True

    # Check vertical
    for col in range(COLS):
        for row in range(ROWS - 3):
            if board[row][col] == piece and board[row+1][col] == piece and board[row+2][col] == piece and board[row+3][col] == piece:
                return True

    # Check positive diagonal
    for col in range(COLS - 3):
        for row in range(ROWS - 3):
            if board[row][col] == piece and board[row+1][col+1] == piece and board[row+2][col+2] == piece and board[row+3][col+3] == piece:
                return True

    # Check negative diagonal
    for col in range(COLS - 3):
        for row in range(3, ROWS):
            if board[row][col] == piece and board[row-1][col+1] == piece and board[row-2][col+2] == piece and board[row-3][col+3] == piece:
                return True

    return False

def get_valid_locations(board):
    valid_locations = []
    for col in range(COLS):
        if is_valid_move(board, col):
            valid_locations.append(col)
    return valid_locations

def Min(current_state, alpha, beta, depth, series_depth, series_count, win_count):
    if is_terminal_node(current_state) or depth == 0:
        return evaluate_board(current_state, PLAYER_X, depth, series_depth, series_count, win_count)

    min_eval = float('inf')

    for next_state in get_next_states(current_state, PLAYER_X):
        eval_val = Max(next_state, alpha, beta, depth - 1, series_depth, series_count, win_count)
        min_eval = min(min_eval, eval_val)
        beta = min(beta, eval_val)
        if beta <= alpha:
            break

    return min_eval

def Max(current_state, alpha, beta, depth, series_depth, series_count, win_count):
    if is_terminal_node(current_state) or depth == 0:
        return evaluate_board(current_state, PLAYER_O, depth, series_depth, series_count, win_count)

    max_eval = float('-inf')

    for next_state in get_next_states(current_state, PLAYER_O):
        eval_val = Min(next_state, alpha, beta, depth - 1, series_depth, series_count, win_count)
        max_eval = max(max_eval, eval_val)
        alpha = max(alpha, eval_val)
        if beta <= alpha:
            break

    return max_eval
    
def get_next_open_row(board, col):
    for r in range(ROWS-1, -1, -1):
        if board[r][col] == EMPTY:
            return r
    

def get_best_move(board, player, depth=3, series_depth=0, series_count=0, win_count=0):
    valid_locations = get_valid_locations(board)
    best_score = float('-inf')
    best_moves = []

    # Pokušaj da pronađemo optimalan potez, izbegavajući otvaranje u ćoškovima
    optimal_moves = list(set(valid_locations) - {0, COLS - 1})

    if len(optimal_moves) > 0:
        best_moves.append(random.choice(optimal_moves))
    else:
        best_moves.append(random.choice(valid_locations))

    for col in valid_locations:
        row = get_next_open_row(board, col)
        temp_board = board.copy()
        drop_piece(temp_board, col, player)
        score = Min(temp_board, float('-inf'), float('inf'), depth, series_depth, series_count, win_count)

        if score > best_score:
            best_moves = [col]
            best_score = score
        elif score == best_score:
            best_moves.append(col)

    # Random izbor između najboljih poteza
    return random.choice(best_moves)


if __name__ == '__main__':
    board = create_board()
    game_over = False
    turn = 0

    while not game_over:
        print(board)
        
        if turn == 0:
            # Bot's turn (Player O)
            col = get_best_move(board, PLAYER_O)
            row = get_next_open_row(board, col)
            drop_piece(board, col, PLAYER_O)
            if winning_move(board, PLAYER_O):
                print(board)
                print("Player O wins!")
                game_over = True
            turn += 1
        else:
            # Human player's turn (Player X)
            col = int(input("Your turn (0-7): "))
            if 0 <= col < COLS and is_valid_move(board, col):
                row = get_next_open_row(board, col)
                drop_piece(board, col, PLAYER_X)
                if winning_move(board, PLAYER_X):
                    print(board)
                    print("You win!")
                    game_over = True
                turn -= 1
            else:
                print("Invalid move! Try again.")

        if len(get_valid_locations(board)) == 0:
            print(board)
            print("It's a tie!")
            game_over = True

